runner/proxy_stream.go

```diff
@@ -2,45 +2,49 @@ package runner
 
 import (
 	"sync"
+	"sync/atomic"
 )
 
 type ProxyStream struct {
-	sync.Mutex
-	subscribers map[int]*Subscriber
-	count       int
+	mu          sync.RWMutex
+	subscribers map[int32]*Subscriber
+	count       atomic.Int32
 }
 
 type Subscriber struct {
-	id       int
+	id       int32
 	reloadCh chan struct{}
 }
 
 func NewProxyStream() *ProxyStream {
-	return &ProxyStream{subscribers: make(map[int]*Subscriber)}
+	return &ProxyStream{subscribers: make(map[int32]*Subscriber)}
 }
 
 func (stream *ProxyStream) Stop() {
 	for id := range stream.subscribers {
 		stream.RemoveSubscriber(id)
 	}
-	stream.count = 0
+	stream.count = atomic.Int32{}
 }
 
 func (stream *ProxyStream) AddSubscriber() *Subscriber {
-	stream.Lock()
-	defer stream.Unlock()
-	stream.count++
+	stream.mu.Lock()
+	defer stream.mu.Unlock()
+	stream.count.Add(1)
 
-	sub := &Subscriber{id: stream.count, reloadCh: make(chan struct{})}
-	stream.subscribers[stream.count] = sub
+	sub := &Subscriber{id: stream.count.Load(), reloadCh: make(chan struct{})}
+	stream.subscribers[stream.count.Load()] = sub
 	return sub
 }
 
-func (stream *ProxyStream) RemoveSubscriber(id int) {
-	stream.Lock()
-	defer stream.Unlock()
-	close(stream.subscribers[id].reloadCh)
-	delete(stream.subscribers, id)
+func (stream *ProxyStream) RemoveSubscriber(id int32) {
+	stream.mu.Lock()
+	defer stream.mu.Unlock()
+
+	if _, ok := stream.subscribers[id]; ok {
+		close(stream.subscribers[id].reloadCh)
+		delete(stream.subscribers, id)
+	}
```

Comment:
need print log when we didn't find subscriber

Reply:
We only have `mainLog` on the Engine struct, we would need to pass it down to Proxy and ProxyStream. I think it's not necessary, I added this just as an extra precaution to avoid nil pointer issues. Missing the subscriber in the map means it's already cleaned up, and during air shutdown we clean up all subscribers, so it's all taken care of. If you still think we should pass down the logger to log here I can do it

Reply:
ic, no need to add